src/hll_hashing.jl

```diff
@@ -0,0 +1,91 @@
+module HyperLogLogHashIPs
+
+export hll_hash_ip, load_hll_key!
+
+using Sockets
+using SHA
+
+function ip_data(addr::IPAddr)
+    n = sizeof(addr.host)
+    [(addr.host >>> ((n-i) << 3)) % UInt8 for i = 1:n]
+end
+
+ip_addr(host::UInt32) = IPv4(host)
+ip_addr(host::UInt128) = IPv6(host)
+
+# Feistel network with sha256 as round function
+
+const ROUNDS = 16
+const DIGEST = SHA.SHA512_CTX
+
+function feistel!(key::Vector{UInt8}, data::Vector{UInt8}, enc::Bool)
+    n = length(data)
+    m = n >> 1
+    rounds = enc ? (0:1:ROUNDS-1) : (ROUNDS-1:-1:0)
+    for round in rounds
+        digest = DIGEST()
+        update!(digest, [round % UInt8])
+        update!(digest, key)
```

Comment:
This is the only use of the key and it doesn't care about how much data is in the key. Its only purpose is to make the Feistel round function hard to predict/invert. In some sense it is choosing a permutation of all possible data values. I.e. for 32-bit data (4 byte), it is choosing one of the `(2^32)! ≈ 2^(2.6e35)` permutations of all the possible 32-bit values. Of course, we can't possibly use a key large enough to access all of those possible permutations — we'd need a key 2.6e35 bits long. But using a key with N bits, for N = 512, 1024 or whatever, is pretty good since it chooses one of 2^N possible permutations of 32-bit values at random and the complex, uninvertible nature of the SHA512 function means that it's basically impossible to understand which permutations are possible.

TLDR: the key can be any data at all, no fixed length.